clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

```diff
@@ -642,6 +660,48 @@ RValue CIRGenFunction::buildBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       Result = builder.createBoolToInt(Result, ResultType);
     return RValue::get(Result);
   }
+
+  case Builtin::BIalloca:
+  case Builtin::BI_alloca:
+  case Builtin::BI__builtin_alloca_uninitialized:
+  case Builtin::BI__builtin_alloca: {
+    // Get alloca size input
+    mlir::Value AllocaSize = buildScalarExpr(E->getArg(0));
+
+    // The alignment of the alloca should correspond to __BIGGEST_ALIGNMENT__.
+    const TargetInfo &TI = getContext().getTargetInfo();
+    const CharUnits AlignmentInBytes =
+        getContext().toCharUnitsFromBits(TI.getSuitableAlign());
+
+    auto AlignIntAttr = CGM.getSize(AlignmentInBytes);
+
+    // Emit the alloca op with type `u8 *` to match the semantics of
+    // `llvm.alloca`. We later bitcast the type to `void *` to match the
+    // semantics of C/C++
+    // FIXME: It may make sense to allow AllocaOp of type `u8` to return a
+    // pointer of type `void *`. This will require a change to the allocaOp
+    // verifier.
+    auto Alloca = builder.create<mlir::cir::AllocaOp>(
+        getLoc(E->getSourceRange()), builder.getUInt8PtrTy(),
+        builder.getUInt8Ty(), "bi_alloca", AlignIntAttr, AllocaSize);
+
+    // Initialize the allocated buffer if required.
+    if (BuiltinID != Builtin::BI__builtin_alloca_uninitialized)
+      initializeAlloca(*this, Alloca, &AllocaSize, AlignmentInBytes);
+
+    // An alloca will always return a pointer to the alloca (stack) address
+    // space. This address space need not be the same as the AST / Language
+    // default (e.g. in C / C++ auto vars are in the generic address space). At
+    // the AST level this is handled within CreateTempAlloca et al., but for the
+    // builtin / dynamic alloca we have to handle it here.
+    // FIXME: introduce proper support for this feature like clang codegen has.
```

Comment:
It seems like address spaces are NYI in CIR. I believe that adding proper support for address spaces in the CIR infrastructure may require deeper thought. I can attempt to tailor exactly what I need for this PR, but perhaps it is a bit out of scope.

Wdyt?

Reply:
What I'm suggesting here is that you add the skeletons to make this look like traditional LLVM codegen, it's fine you might not use this for now. You should also add `assert(!UnimplementedFeature::addressSpace());` here.

Reply:
Since `getASTAllocaAddressSpace()` is already declared as an `UnimplementedFeature`, and since - not coincidentally - `createTempAlloca` relies on it staying unimplemented, I would want to suggest again to delay this to future PR. I will obviously add both asserts here to make sure this code is fixed once address spaces are supported.

Another option would be to implement it but to always return the default address space, and assert in this new implementation that addressSpace handling is not yet supported.

Do you have a preferred path forward? Or perhaps a third alternative?

Reply:
Would adding the following to `TargetCIRGenInfo` make sense to you?
```
  /// Get the AST address space for alloca.
  virtual LangAS getASTAllocaAddressSpace() const {
    // Address spaces are not yet fully supported, but the usage of the default
    // alloca address space can be used for now only for comparison with the
    // default address space.
    assert(!UnimplementedFeature::addressSpace);
    return LangAS::Default;
  }
```

This seems reasonable, but has potential to blow up in someone's face if used improperly.

Reply:
> Since getASTAllocaAddressSpace() is already declared as an UnimplementedFeature, and since - not coincidentally - createTempAlloca relies on it staying unimplemented, I would want to suggest again to delay this to future PR. I will obviously add both asserts here to make sure this code is fixed once address spaces are supported.

I missed that, good point.

> Another option would be to implement it but to always return the default address space, and assert in this new implementation that addressSpace handling is not yet supported.

That'd be my ideal option, so we can probably diminish the amount of places we'll need to add such assert, luckily they will be all abstracted by `getASTAllocaAddressSpace`.

> Would adding the following to TargetCIRGenInfo make sense to you?

You should add the assert to `CIRGenTypeCache::getASTAllocaAddressSpace` (plz add that one for consistency). For  `TargetCIRGenInfo::getASTAllocaAddressSpace` you don't need the assert, because specific behavior needs to be added by targets anyways, and the default method in LLVM traditional codegen just simply returns `LangAS::Default` too.

Reply:
Great, just applied a review there too!